src/Content/Text/BBCode.php

```diff
@@ -1341,581 +1339,611 @@ public static function convert(string $text = null, bool $try_oembed = true, int
+						$heading--;
+					}
 				}
+			}
+		} else {
+			$text = preg_replace("(\[h1\](.*?)\[\/h1\])ism", '</p><h1>$1</h1><p>', $text);
+			$text = preg_replace("(\[h2\](.*?)\[\/h2\])ism", '</p><h2>$1</h2><p>', $text);
+			$text = preg_replace("(\[h3\](.*?)\[\/h3\])ism", '</p><h3>$1</h3><p>', $text);
+			$text = preg_replace("(\[h4\](.*?)\[\/h4\])ism", '</p><h4>$1</h4><p>', $text);
+			$text = preg_replace("(\[h5\](.*?)\[\/h5\])ism", '</p><h5>$1</h5><p>', $text);
+			$text = preg_replace("(\[h6\](.*?)\[\/h6\])ism", '</p><h6>$1</h6><p>', $text);
+		}
 
-				// Check for list text
-				$text = str_replace("[*]", "<li>", $text);
-				$text = str_replace("[li]", "<li>", $text);
-
-				$text = preg_replace("/\[th\](.*?)\[\/th\]/sm", '<th>$1</th>', $text);
-				$text = preg_replace("/\[td\](.*?)\[\/td\]/sm", '<td>$1</td>', $text);
-				$text = preg_replace("/\[tr\](.*?)\[\/tr\]/sm", '<tr>$1</tr>', $text);
-				$text = preg_replace("/\[table\](.*?)\[\/table\]/sm", '</p><table>$1</table><p>', $text);
-
-				$text = preg_replace("/\[table border=1\](.*?)\[\/table\]/sm", '</p><table border="1" >$1</table><p>', $text);
-				$text = preg_replace("/\[table border=0\](.*?)\[\/table\]/sm", '</p><table border="0" >$1</table><p>', $text);
-
-				$text = str_replace('[hr]', '</p><hr /><p>', $text);
-
-				if (!$for_plaintext) {
-					$text = self::performWithEscapedTags($text, ['url', 'img', 'audio', 'video', 'youtube', 'vimeo', 'share', 'attachment', 'iframe', 'bookmark'], function ($text) {
-						return preg_replace(Strings::autoLinkRegEx(), '[url]$1[/url]', $text);
-					});
-				}
+		return $text;
+	}
 
-				// Check for font change text
-				$text = preg_replace("/\[font=(.*?)\](.*?)\[\/font\]/sm", "<span style=\"font-family: $1;\">$2</span>", $text);
+	private static function convertEmojisToHtml(string $text, int $simple_html): string
+	{
+		// Mastodon Emoji (internal tag, do not document for users)
+		if ($simple_html == self::MASTODON_API) {
+			$text = preg_replace("(\[emoji=(.*?)](.*?)\[/emoji])ism", '$2', $text);
+		} else {
+			$text = preg_replace("(\[emoji=(.*?)](.*?)\[/emoji])ism", '<span class="mastodon emoji"><img src="$1" alt="$2" title="$2"/></span>', $text);
+		}
+		return $text;
+	}
 
-				// Declare the format for [spoiler] layout
-				$SpoilerLayout = '<details class="spoiler"><summary>' . DI::l10n()->t('Click to open/close') . '</summary>$1</details>';
+	private static function convertStylesToHtml(string $text, int $simple_html): string
+	{
+		// Several easy to replace HTML elements
+		// @todo add the new elements to the documentation by the end of 2024 so that most systems will support them.
+		$elements = ['b', 'del', 'em', 'i', 'ins', 'kbd', 'mark',
+			's', 'samp', 'small', 'strong', 'sub', 'sup', 'u', 'var'];
+		foreach ($elements as $element) {
+			$text = preg_replace("(\[" . $element . "\](.*?)\[\/" . $element . "\])ism", '<' . $element . '>$1</' . $element . '>', $text);
+		}
```

Comment:
I don't think the `[i]` and `[em]` tags should convert to different HTML tags because we don't differentiate their use in the BBCode documentation. As a result, I believe we should be able to specify an alternative HTML tag a BBCode should convert to.
```suggestion
		$elements = ['b', 'del' => 's', 'em' => 'i', 'i', 'ins', 'kbd', 'mark',
			's', 'samp', 'small', 'strong' => 'b', 'sub', 'sup', 'u', 'var'];
		foreach ($elements as $source => $element) {
			$source = is_string($source) ? $source : $element;
			$text = preg_replace("(\[" . $source . "\](.*?)\[\/" . $source . "\])ism", '<' . $element . '>$1</' . $element . '>', $text);
		}
```

Reply:
"i" and "em" are different tags, although they look the same. The same goes with "del" and "s". Especially for this element we could use some changed CSS for this element.

Reply:
We could but we don't at the moment, and we're presenting formatting options rather than semantic elements in the post compose form, so I believe we should stick to the corresponding HTML tags `<i>` and `<s>`.

Reply:
I would like to use some CSS like the one used here https://developer.mozilla.org/en-US/docs/Web/HTML/Element/del or here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ins in the future.

Reply:
That's cool, but we can't assume remote systems will have adequate support for any of them, so I believe we should stick to the lowest common denominator tags, `<i>` and `<s>`. I guess that's the logical downside of trying to be compatible with as many different protocols as possible.

Reply:
We can import web pages via the feed functionality. BTW: I tested all these elements with Mastodon and Pleroma. Pleroma for example doesn't render `s`, but `del`.

Friendica:
![image](https://github.com/friendica/friendica/assets/844208/cb99242c-f524-4763-8ede-7f5ef9002d66)

Pleroma:
![image](https://github.com/friendica/friendica/assets/844208/4e59c2c4-bbdc-4df8-af8b-96e0a4d93777)

We can't assume, that remote systems will render elements that look identical on Friendica will look identical on their system. So we shouldn't assume that.